tests/unit/test_errors.py

```diff
@@ -0,0 +1,117 @@
+# Copyright 2021 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+import subprocess
+
+from craft_providers import errors
+
+
+def test_provider_error():
+    error = errors.ProviderError(brief="test brief")
+
+    assert error == errors.ProviderError(brief="test brief")
+    assert str(error) == "test brief"
+
+
+def test_provider_error_with_details():
+    error = errors.ProviderError(brief="test brief", details="test details")
+
+    assert error == errors.ProviderError(brief="test brief", details="test details")
+    assert str(error) == "test brief\ntest details"
+
+
+def test_provider_error_with_resolution():
+    error = errors.ProviderError(brief="test brief", resolution="test resolution")
+
+    assert error == errors.ProviderError(
+        brief="test brief", resolution="test resolution"
+    )
+    assert str(error) == "test brief\ntest resolution"
+
+
+def test_provider_error_with_all():
+    error = errors.ProviderError(
+        brief="test brief", details="test details", resolution="test resolution"
+    )
+
+    assert error == errors.ProviderError(
+        brief="test brief", details="test details", resolution="test resolution"
+    )
+    assert str(error) == "test brief\ntest details\ntest resolution"
+
+
+def test_details_from_called_process_error():
+    error = subprocess.CalledProcessError(
+        -1, ["test-command", "flags", "quote$me"], "test stdout", "test stderr"
+    )
+
+    details = errors.details_from_called_process_error(error)
+
+    assert details == "\n".join(
```

Comment:
why not textwrap dedent instead?

Reply:
I'll update to use it.

Reply:
done.

Reply:
(also added a decode() behavior for bytes with additional test case)